adapters/insticator/insticator.go

```diff
@@ -0,0 +1,526 @@
+
+// Map parameters to validation functions
+var optionalVideoParams = map[string]func(interface{}) bool{
+	"minduration":    isInteger,
+	"maxduration":    isInteger,
+	"protocols":      isArrayOfNums,
+	"startdelay":     isInteger,
+	"linearity":      func(value interface{}) bool { return isInteger(value) && validLinearity[toInt(value)] },
+	"skip":           func(value interface{}) bool { return isInteger(value) && validSkip[toInt(value)] },
+	"skipmin":        isInteger,
+	"skipafter":      isInteger,
+	"sequence":       isInteger,
+	"battr":          isArrayOfNums,
+	"maxextended":    isInteger,
+	"minbitrate":     isInteger,
+	"maxbitrate":     isInteger,
+	"playbackmethod": isArrayOfNums,
+	"playbackend":    func(value interface{}) bool { return isInteger(value) && validPlaybackEnd[toInt(value)] },
+	"delivery":       isArrayOfNums,
+	"pos":            func(value interface{}) bool { return isInteger(value) && validPos[toInt(value)] },
+	"api":            isArrayOfNums,
+}
+
+// Helper function to convert interface to int
+func toInt(value interface{}) int {
+	switch v := value.(type) {
+	case int:
+		return v
+	case float64:
+		return int(v)
+	case string:
+		if i, err := strconv.Atoi(v); err == nil {
+			return i
+		}
+	}
+	return 0
+}
+
+// Builder builds a new insticatorance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// getMediaTypeForImp figures out which media type this bid is for
+func getMediaTypeForBid(bid *openrtb2.Bid) openrtb_ext.BidType {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo
+	default:
+		return openrtb_ext.BidTypeBanner
+	}
+}
+
+func getBidType(ext bidExt) openrtb_ext.BidType {
+	if ext.Insticator.MediaType == "video" {
+		return openrtb_ext.BidTypeVideo
+	}
+
+	return openrtb_ext.BidTypeBanner
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+	var groupedImps = make(map[string][]openrtb2.Imp)
+
+	reqExt, err := makeReqExt(request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	request.Ext = reqExt
+
+	for i := 0; i < len(request.Imp); i++ {
+		if impCopy, err := makeImps(request.Imp[i]); err == nil {
+			var impExt Ext
+			// Populate site.publisher.id from imp extension only once
+			if request.Site != nil && i == 0 {
+				populatePublisherId(&impCopy, request)
+			} else if request.App != nil && i == 0 {
+				populatePublisherId(&impCopy, request)
+			}
+
+			// group together the imp hacing insticator adUnitId. However let's not block request creation.
+			if err := json.Unmarshal(impCopy.Ext, &impExt); err == nil {
```

Comment:
generally, error not nil  is checked first and then check is followed by other logic when error is nil
should follow same pattern here

```
if err := json.Unmarshal(impCopy.Ext, &impExt); err !=nil {
  errs = append(errs, err)
  continue
}

// logic when `err` is nil
```